### PR TITLE
Fixing issue with ZEPPELIN-173: Zeppelin websocket server is vulnerab…

### PR DESCRIPTION
**Pull request by djoelz**
Fixing the socket cross-origin vulnerability as described in the Jira. Overwrote the checkOrigin in the WebSocketServlet class implemented by NotebookServer so that a list of all seen socket Get requests are kept and only Upgrade requests from the same origin will be accepted. Otherwise unauthorized will be returned.
Included basic unit tests.

### CODE REVIEW

File: zeppelin-server/src/main/java/org/apache/zeppelin/socket/NotebookServer.java [hunk 1]
```diff
@@ -333,20 +324,13 @@ private boolean isCronUpdated(Map<String, Object> configA,
     }
     return cronUpdated;
   }
-
-  private void createNote(WebSocket conn, Notebook notebook, Message message) throws IOException {
+  private void createNote(WebSocket conn, Notebook notebook) throws IOException {
     Note note = notebook.createNote();
     note.addParagraph(); // it's an empty note. so add one paragraph
-    if (message != null) {
```
Comment (Leemoonsoo) [inline zeppelin-server/src/main/java/org/apache/zeppelin/socket/NotebookServer.java]: Could you explain little bit, why this if condition can be removed??

Comment (djoelz) [inline zeppelin-server/src/main/java/org/apache/zeppelin/socket/NotebookServer.java]: I think it's a merge issue as I did not remove it. Let me double check

**Comment by Leemoonsoo** [comment]
Tested and working nicely. Thanks for the contribution!

File: zeppelin-server/src/main/java/org/apache/zeppelin/socket/NotebookServer.java (hunk 1 above)
Comment (djoelz) [inline zeppelin-server/src/main/java/org/apache/zeppelin/socket/NotebookServer.java]: Fixing in the next commit. Please review when you have time.

**Comment by djoelz** [comment]
I have fixed the merge issues and recommitted. Ready for your review.

Thanks,
Joel
________________________________

**Comment by Leemoonsoo** [comment]
Thanks, LGTM.

**Comment by djoelz** [comment]
Great! Next step is to merge? Who does this?

**Comment by Leemoonsoo** [comment]
Next step is, getting more review and votes, or waiting for enough time to have discussions and consensus (which is normally take a day at least). Then it's going to be merged.

**Comment by philwills** [comment]
Not sure if it's best to comment here, or open a new issue, but 

```java
java.net.InetAddress.getLocalHost().getHostName();
```

isn't going to return all possible addresses which a node might reasonably be listening on. For instance, on an EC2 node, this will return the private IP, but if you want to connect to that node from outside of Amazon's network, that address won't be visible, where as the public address will.

I think there needs to at least be the option of setting an alternative value in config.

**Comment by corneadoug** [comment]
We got a similar problem, can't complete websocket handshake in some instalations since this commit

**Comment by Leemoonsoo** [comment]
@philwills @corneadoug Right, i'll create a patch, soon. Thanks!

**Comment by djoelz** [comment]
@Leemoonsoo can I suggest alternativeallowedsource as the configuration name? Also this will be used for my other pull request that affects REST endpoints as well.

I could implement it also if you want. have you started already?

**Comment by Leemoonsoo** [comment]
@djoelz If you can implement, that would be really appreciated!

**Comment by djoelz** [comment]
@Leemoonsoo @jonbuffington is already doing the work. I will work closely with Jon to wrap this up.

Thanks Jon!

**Comment by djoelz** [comment]
I have a fix for this. Will create the PR soon

**Comment by jitenderaswani** [comment]
Looking forward to this fix, I am unable to run Zeppelin in AWS.  On my local machine, I don't have web-socket issue.

**Comment by Leemoonsoo** [comment]
@djoelz I also pushed a fix. @djoelz, @jitenderaswani  please review #233.

**Pull request closed without merging**